### PR TITLE
View smart search

### LINKED ISSUE #474: Create/edit Smart Search for View
## Description
A user should be able to configure a Smart Search query in a view, so that the view is dynamically populated with the people matching that Smart Search query. When the result of the query changes, so does the content of the view.

This is different from a "static" view, where people are added one by one to the view (or from another view, as described in #476), and thus, the user needs to be able to pick whether the view should be static or dynamic.

## Relevant Job Stories
* When I'm brainstorming people to add to my view as a shortlist, I want to be able to define a Smart Search query and see the resulting data set in my view, so that I can add the ones I want to the view from there
* When I'm using a Smart Search query to populate my view, I want to be able to save it in the view so that every time I come back the view contains the people currently matching that query

## Prerequisites
This task requires changes to the Views API. See updates in comments below, or ask if unsure.

## Requirements
- [x] On the single-view page, if a view is empty, it should ask the user to decide whether to statically add specific people or define a Smart Search
- [x] If user picks static, render the empty table and the line where user can add records (#475)
- [x] If user picks Smart Search, open the Smart Search edit modal, and save the Smart Search to the view

## Possible implementations
Use the existing Smart Search modal to define Smart Search. Attach the query to the view using `PATCH /orgs/{org_id}/people/views/{view_id}` (pending changes to API).

## Design specifications
image

## Open questions
No outstanding questions at this time.

### PR DESCRIPTION
**Pull request by richardolsson**
## Description
This PR implements a UI for empty views, which allows the user to add the first person ("static") or, more interestingly, configure a Smart Search query for the view ("dynamic"). It also adds an "ellipsis menu" that can be used to convert between static and dynamic views.

## Screenshots
image

image

## Changes
* Refactor SmartSearchDialog to work in any context (and add tests)
* Refactor PersonSelect to work in any context (and add tests)
* Add general-purpose solution for "ellipsis menus" in `TabbedLayout`
* Implement UI for empty views
  * Card where user can pick a person to add as first row
  * Card where user can trigger `SmartSearchDialog` and configure a query
  * Menu to edit Smart Search query or convert between query/no query

See #474 for more details on the feature.

## Notes to reviewer
The View `content_query` feature is in development in the Zetkin Platform API and has not yet made it onto the live dev server. This means that the Smart Search feature will not work unless you're able to run a development version of the Zetkin Platform. Adding a person manually should still work though.

Please focus on reviewing the code for now. Reach out to me for a demo.

## Related issues
Resolves #474

### CODE REVIEW

File: playwright/tests/organize/campaign-detail/action-buttons.spec.ts
Comment (djbusstop) [inline playwright/tests/organize/campaign-detail/action-buttons.spec.ts:90]: Thanks for adding these, was recently thinking about how we weren't testing requests when we built this.

File: playwright/tests/organize/campaign-detail/action-buttons.spec.ts
Comment (djbusstop) [inline playwright/tests/organize/campaign-detail/action-buttons.spec.ts:87]: This doesn't need await right? Also, you can destructure with `const { log } = await moxy.logRequests()

File: playwright/tests/organize/people/views/view-detail.spec.ts
Comment (djbusstop) [inline playwright/tests/organize/people/views/view-detail.spec.ts:178]: beautiful tests

File: src/components/forms/CampaignDetailsForm.tsx
Comment (djbusstop) [inline src/components/forms/CampaignDetailsForm.tsx:97]: Hell yes `<PersonSelect />

File: src/components/forms/common/PersonSelect.tsx
Comment (djbusstop) [inline src/components/forms/common/PersonSelect.tsx:16]: I think this makes more sense to just get internally using `router.query`, since if this moves in to a component without the `orgId`, then you have to add that prop to the parent OR do this anyway.

File: src/components/forms/common/PersonSelect.tsx
Comment (djbusstop) [inline src/components/forms/common/PersonSelect.tsx:73]: very small thing but one letter variable names leave me sitting there going "what does v stand for". i realised after far too long that it's value. I would prefer more clear names. The one place I'm used to a one letter name is event callbacks having `e` as an argument, because that's a common pattern. Don't have to change if you don't want to but `e, value` would confuse me less.

File: src/components/layout/organize/SingleViewLayout.tsx
Comment (djbusstop) [inline src/components/layout/organize/SingleViewLayout.tsx:69]: Maybe call this `ellipsisMenuItems`?

File: src/components/layout/organize/TabbedLayout.tsx
Comment (djbusstop) [inline src/components/layout/organize/TabbedLayout.tsx:123]: I don't love that we now have two implementations of the same thing here, since the action buttons hold the existing menues. I don't know if we should solve it in this PR... I guess the solution isn't to remove action buttons, because they could still be used for the high impact buttons in the title area. But that we should move all existing menues to this method.

Comment (richardolsson) [reply src/components/layout/organize/TabbedLayout.tsx:123]: I agree, but I figured it was outside the scope of this PR. I was thinking we could create separate issues, or even defer such changes until the next time we actually work on those menus. I should have communicated that somehow.

Maybe we should add a section to the PR template called "Deferred tasks" or "Out of scope" or something?

**Review by djbusstop — COMMENTED** [review]
Some small comments, nothing major! If you don't want to change what I suggested, it's not blocking

File: src/components/layout/organize/TabbedLayout.tsx
Comment (djbusstop) [reply src/components/layout/organize/TabbedLayout.tsx:123]: Yea maybe we should...

**Comment by richardolsson** [comment]
> Some small comments, nothing major! If you don't want to change what I suggested, it's not blocking

Great review! I agree with all of it and will make the necessary changes (as well as fix the merge conflict). 🙌

File: src/components/forms/common/PersonSelect.tsx
Comment (richardolsson) [reply src/components/forms/common/PersonSelect.tsx:16]: I guess you're right! I'm not yet comfortable with this approach to objects retrieving their dependencies rather than having them injected, and I do think it complicates testing (in this case `useRouter()` will need to be mocked to test `PersonSelect`) but I'm willing to keep trying and hopefully will feel comfortable soon enough. 😊

File: src/components/forms/common/PersonSelect.tsx
Comment (richardolsson) [reply src/components/forms/common/PersonSelect.tsx:73]: I agree! This was probably just carried across from the previous implementation inside `CampaignDetailsForm`. I'm renaming them, but prefer to use `ev` for events (to not confuse them with errors, which `e` could also be read as).

File: playwright/tests/organize/campaign-detail/action-buttons.spec.ts
Comment (richardolsson) [reply playwright/tests/organize/campaign-detail/action-buttons.spec.ts:87]: Tru dat!

File: src/components/layout/organize/SingleViewLayout.tsx
Comment (richardolsson) [reply src/components/layout/organize/SingleViewLayout.tsx:69]: Good point!

File: src/components/layout/organize/TabbedLayout.tsx
Comment (richardolsson) [reply src/components/layout/organize/TabbedLayout.tsx:123]: I have created separate issue #522 for this

**Comment by richardolsson** [comment]
Outside your window snow is falling
This relaxing Christmas day
When your Thunderbird comes calling
It is GitHub saying "Hey"
But just ignore it, it's not pressing
I'm just being COVID bored
Instead of work, make sure you're resting
With Santa, commies and the Lord

But yeah, maybe have a look at this Monday, @djbusstop?

**Comment by richardolsson** [comment]
I crammed in a refactor of the `PersonSelect`, moving all the logic out into a hook and then using that for two variants of the component, one `mui-rff` compatible and one basic MUI `Autocomplete` implementation.

File: src/components/forms/common/PersonSelect.tsx
Comment (djbusstop) [reply src/components/forms/common/PersonSelect.tsx]: Actually we don't need to mock `useRouter` im pretty sure, what you do to set values in the router is to use this pattern:

```js
    singletonRouter.query = {
        orgId: '1',
    };
```

We did this in `TaskStatusSubtitle.spec.tsx` if you wanna see it in use

**Review by djbusstop — DISMISSED** [review]
All good. Great work.

File: src/components/forms/common/PersonSelect.tsx
Comment (richardolsson) [reply src/components/forms/common/PersonSelect.tsx]: Yeah I realized that testing isn't that big of a deal. But I did find another scenario where this was causing extra work, my view thumbnail prototype. But there are always going to be pros and cons that outweigh each other in different situations, so I don't mind!

**Comment by richardolsson** [comment]
Thanks! 😊 

I'm going to hold off on this until the necessary API changes have been merged, and also there's a merge conflict here now that I'll take care of when it's ready to merge.

**Comment by richardolsson** [comment]
Do you mind doing a final review @djbusstop? I've deployed the API changes to the platform dev server and solved the merge conflicts with `main` now. This means you can actually try the feature.

**Review by djbusstop — APPROVED** [review]

**Pull request merged**
